File: solutions/python/code-carl/ch09_dp/t10_lc1049_last_stone_weight_ii.py
"""LeetCode 1049. Last Stone Weight II
https://leetcode.cn/problems/last-stone-weight-ii/
"""

from typing import List


class Solution:
    def lastStoneWeightII(self, stones: List[int]) -> int:
        total = sum(stones)
        bag_size = total // 2
        dp: List[int] = [0] * (bag_size + 1)

        for w in stones:
            for j in range(bag_size, w - 1, -1):
                dp[j] = max(dp[j], dp[j - w] + w)

        return total - 2 * dp[-1]


if __name__ == "__main__":
    solution = Solution()
    assert solution.lastStoneWeightII([2, 7, 4, 1, 8, 1]) == 1
    assert solution.lastStoneWeightII([31, 26, 33, 21, 40]) == 5
    print("All tests passed.")

File: solutions/code-carl/ch09_dp/T10_LC1049_LastStoneWeightIi.java
/**
 * Practice template for:
 * - LeetCode 1049. Last Stone Weight II
 *   https://leetcode.cn/problems/last-stone-weight-ii/
 */
/**
 * @author Cheng Yihao
 * @version 1.0
 * @date 2024/11/21 19:31
 * @comment
 */
public class T10_LC1049_LastStoneWeightIi {
    public int lastStoneWeightII(int[] stones) {
        int sum = 0;
        for (int n : stones) {
            sum += n;
        }

        int bagSize = sum / 2;
        int[] dp = new int[bagSize + 1];
        for (int i = 0; i < stones.length; i++) {
            for (int j = bagSize; j >= stones[i]; j--) {
                dp[j] = Math.max(dp[j], dp[j - stones[i]] + stones[i]);
            }
        }
        return sum - 2 * dp[bagSize];
    }
}
